File: esbmc.cpp/validador_categoria/validador_final_categoria.cpp
#include <cctype>

#define MAX_LEN 31

using namespace std;

bool validate_no_end_special_or_digit(const char category[MAX_LEN], int len) {
    const char specials[] = "!@#$%^&*()_+=-[]{}|;:,.<>?/~`";
    char last = category[len - 1];
    for (int j = 0; specials[j] != '\0'; j++) {
        __ESBMC_assert(last != specials[j], "A categoria não pode terminar com um número ou caractere especial.");
    }
    __ESBMC_assert(!isdigit(last), "A categoria não pode terminar com um número ou caractere especial.");
    return true;
}

int main() {
    char category[MAX_LEN];
    int len;

    __ESBMC_assume(len >= 3 && len <= 30);

    validate_no_end_special_or_digit(category, len);

    return 0;
}

File: esbmc.py/Testes02/validador_categoria/validador_final_categoria.py
MAX_LEN: int = 31

def validate_no_end_special_or_digit(category: str, length: int) -> bool:
    specials: str = "!@#$%^&*()_+=-[]{}|;:,.<>?/~`"
    last: str = category[length - 1]

    j: int = 0
    while j < len(specials):
        assert last != specials[j], \
            "A categoria não pode terminar com um número ou caractere especial."
        j += 1

    assert not last.isdigit(), \
        "A categoria não pode terminar com um número ou caractere especial."

    return True


def main() -> None:
    category: str = input()  
    length: int = len(category)

    assert length >= 3 and length <= 30

    validate_no_end_special_or_digit(category, length)


main()
